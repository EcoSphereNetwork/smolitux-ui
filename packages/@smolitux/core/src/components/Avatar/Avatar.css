.smolitux-avatar {
  position: relative;
  display: inline-flex;
  align-items: center;
  justify-content: center;
  flex-shrink: 0;
  background-color: var(--avatar-bg, #e5e7eb);
  color: var(--avatar-text-color, #1f2937);
  font-weight: 500;
  overflow: hidden;
}

.smolitux-avatar-image {
  width: 100%;
  height: 100%;
  object-fit: cover;
}

.smolitux-avatar-initials {
  text-transform: uppercase;
  line-height: 1;
}

/* Größen */
.smolitux-avatar--xs {
  width: 1.5rem;
  height: 1.5rem;
  font-size: 0.625rem;
}

.smolitux-avatar--sm {
  width: 2rem;
  height: 2rem;
  font-size: 0.75rem;
}

.smolitux-avatar--md {
  width: 2.5rem;
  height: 2.5rem;
  font-size: 0.875rem;
}

.smolitux-avatar--lg {
  width: 3rem;
  height: 3rem;
  font-size: 1rem;
}

.smolitux-avatar--xl {
  width: 4rem;
  height: 4rem;
  font-size: 1.25rem;
}

/* Formen */
.smolitux-avatar--circle {
  border-radius: 50%;
}

.smolitux-avatar--square {
  border-radius: 0;
}

.smolitux-avatar--rounded {
  border-radius: 0.375rem;
}

/* Rahmen */
.smolitux-avatar--bordered {
  border: 2px solid var(--avatar-border-color, #e5e7eb);
}

.smolitux-avatar--border-primary {
  border-color: var(--primary-500, #3b82f6);
}

.smolitux-avatar--border-secondary {
  border-color: var(--secondary-500, #8b5cf6);
}

.smolitux-avatar--border-success {
  border-color: var(--green-500, #10b981);
}

.smolitux-avatar--border-warning {
  border-color: var(--yellow-500, #f59e0b);
}

.smolitux-avatar--border-error {
  border-color: var(--red-500, #ef4444);
}

.smolitux-avatar--border-info {
  border-color: var(--blue-500, #3b82f6);
}

.smolitux-avatar--border-gray {
  border-color: var(--gray-300, #d1d5db);
}

/* Status */
.smolitux-avatar-status {
  position: absolute;
  bottom: 0;
  right: 0;
  width: 0.75rem;
  height: 0.75rem;
  border-radius: 50%;
  border: 2px solid var(--avatar-status-border-color, white);
}

.smolitux-avatar--xs .smolitux-avatar-status,
.smolitux-avatar--sm .smolitux-avatar-status {
  width: 0.5rem;
  height: 0.5rem;
  border-width: 1px;
}

.smolitux-avatar--lg .smolitux-avatar-status,
.smolitux-avatar--xl .smolitux-avatar-status {
  width: 1rem;
  height: 1rem;
  border-width: 2px;
}

.smolitux-avatar-status--online {
  background-color: var(--avatar-status-online-color, #10b981);
}

.smolitux-avatar-status--offline {
  background-color: var(--avatar-status-offline-color, #6b7280);
}

.smolitux-avatar-status--away {
  background-color: var(--avatar-status-away-color, #f59e0b);
}

.smolitux-avatar-status--busy {
  background-color: var(--avatar-status-busy-color, #ef4444);
}

/* Avatar-Gruppe */
.smolitux-avatar-group {
  display: inline-flex;
  align-items: center;
}

.smolitux-avatar-group .smolitux-avatar {
  border: 2px solid var(--avatar-group-border-color, white);
  margin-right: var(--avatar-group-spacing, -0.5rem);
}

.smolitux-avatar-group .smolitux-avatar:hover {
  transform: translateY(-0.25rem);
  transition: transform 0.2s ease-in-out;
  z-index: 1;
}

/* Avatar mit Anzahl */
.smolitux-avatar-count {
  background-color: var(--avatar-count-bg, #e5e7eb);
  color: var(--avatar-count-color, #1f2937);
  font-size: 0.75em;
}

/* Fallback-Animation */
@keyframes smolitux-avatar-pulse {
  0% {
    opacity: 1;
  }
  50% {
    opacity: 0.7;
  }
  100% {
    opacity: 1;
  }
}

.smolitux-avatar--loading {
  animation: smolitux-avatar-pulse 1.5s infinite;
}

/* Dark Mode */
.dark .smolitux-avatar {
  background-color: var(--avatar-bg-dark, #4b5563);
  color: var(--avatar-text-color-dark, #f9fafb);
}

.dark .smolitux-avatar--bordered {
  border-color: var(--avatar-border-color-dark, #374151);
}

.dark .smolitux-avatar--border-gray {
  border-color: var(--gray-600, #4b5563);
}

.dark .smolitux-avatar-status {
  border-color: var(--avatar-status-border-color-dark, #1f2937);
}

.dark .smolitux-avatar-group .smolitux-avatar {
  border-color: var(--avatar-group-border-color-dark, #1f2937);
}

.dark .smolitux-avatar-count {
  background-color: var(--avatar-count-bg-dark, #4b5563);
  color: var(--avatar-count-color-dark, #f9fafb);
}