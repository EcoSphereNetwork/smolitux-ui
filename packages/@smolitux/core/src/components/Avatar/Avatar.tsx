// packages/@smolitux/core/src/components/Avatar/Avatar.tsx
import React, { useState } from 'react';

export interface AvatarProps extends Omit<React.HTMLAttributes<HTMLDivElement>, 'color'> {
  /** Bildquelle URL */
  src?: string;
  /** Ausweichtext, wenn Bild nicht geladen werden kann */
  alt?: string;
  /** Größe des Avatars */
  size?: 'xs' | 'sm' | 'md' | 'lg' | 'xl';
  /** Platzhaltername (für Avatar ohne Bild) */
  name?: string;
  /** Rahmen hinzufügen */
  bordered?: boolean;
  /** Rahmenfarbe */
  borderColor?: 'primary' | 'secondary' | 'success' | 'warning' | 'error' | 'info' | 'gray' | string;
  /** Status */
  status?: 'online' | 'offline' | 'away' | 'busy';
  /** Benutzerdefinierte Komponente */
  customComponent?: React.ReactNode;
  /** Form des Avatars */
  shape?: 'circle' | 'square' | 'rounded';
  /** Gruppe von Avataren */
  group?: boolean;
  /** Position im Stack (für Gruppen) */
  stackIndex?: number;
}

/**
 * Avatar-Komponente für Benutzer oder Profilbilder
 * 
 * @example
 * ```tsx
 * <Avatar src="/images/profile.jpg" alt="Profilbild" size="md" status="online" />
 * ```
 */
export const Avatar: React.FC<AvatarProps> = ({
  src,
  alt = 'Avatar',
  size = 'md',
  name,
  bordered = false,
  borderColor = 'gray',
  status,
  customComponent,
  shape = 'circle',
  group = false,
  stackIndex = 0,
  className = '',
  ...rest
}) => {
  // State für Bildladefehler
  const [imgError, setImgError] = useState(false);
  
  // Größen-spezifische Klassen
  const sizeClasses = {
    xs: 'h-6 w-6 text-xs',
    sm: 'h-8 w-8 text-sm',
    md: 'h-12 w-12 text-base',
    lg: 'h-16 w-16 text-lg',
    xl: 'h-20 w-20 text-xl'
  };
  
  // Status-spezifische Klassen
  const statusClasses = {
    online: 'bg-green-500',
    offline: 'bg-gray-400',
    away: 'bg-yellow-500',
    busy: 'bg-red-500'
  };
  
  // Form-spezifische Klassen
  const shapeClasses = {
    circle: 'rounded-full',
    square: 'rounded-none',
    rounded: 'rounded-md'
  };
  
  // Gruppen-spezifische Klassen und Stile
  const groupStyles = group ? {
    marginLeft: stackIndex > 0 ? `-${stackIndex * 8}px` : '0',
    zIndex: 10 - stackIndex
  } : {};
  
  // Border-Farben
  const getBorderColorClass = () => {
    // Predefined colors
    const colorMap: Record<string, string> = {
      primary: 'border-primary-500 dark:border-primary-600',
      secondary: 'border-secondary-500 dark:border-secondary-600',
      success: 'border-green-500 dark:border-green-600',
      warning: 'border-yellow-500 dark:border-yellow-600',
      error: 'border-red-500 dark:border-red-600',
      info: 'border-blue-500 dark:border-blue-600',
      gray: 'border-gray-200 dark:border-gray-700'
    };
    
    return colorMap[borderColor] || `border-${borderColor}`;
  };
  
  // Generiere Initialen aus dem Namen
  const getInitials = () => {
    if (!name) return '';
    
    const names = name.split(' ');
    if (names.length === 1) return names[0].charAt(0).toUpperCase();
    
    return (names[0].charAt(0) + names[names.length - 1].charAt(0)).toUpperCase();
  };
  
  // Bestimme Hintergrundfarbe für Platzhalter basierend auf dem Namen
  const getBackgroundColor = () => {
    if (!name) return 'bg-gray-300 dark:bg-gray-600';
    
    // Einfache Hash-Funktion für Namen
    const hash = name.split('').reduce((acc, char) => {
      return acc + char.charCodeAt(0);
    }, 0);
    
    // Liste von Farben
    const colors = [
      'bg-red-500', 'bg-blue-500', 'bg-green-500', 
      'bg-yellow-500', 'bg-purple-500', 'bg-pink-500',
      'bg-indigo-500', 'bg-teal-500'
    ];
    
    return colors[hash % colors.length];
  };
  
  // Basis-Klassen für den Avatar
  const avatarClasses = [
    'relative inline-flex items-center justify-center overflow-hidden',
    sizeClasses[size],
    shapeClasses[shape],
    bordered ? `border-2 ${getBorderColorClass()}` : '',
    'bg-gray-200 dark:bg-gray-700',
    group ? 'ring-2 ring-white dark:ring-gray-800' : '',
    className
  ].join(' ');
  
  return (
    <div className={avatarClasses} style={groupStyles} {...rest}>
      {/* Custom Component */}
      {customComponent && (
        <div className="w-full h-full">{customComponent}</div>
      )}
      
      {/* Bild */}
      {!customComponent && src && !imgError && (
        <img 
          src={src} 
          alt={alt} 
          className="h-full w-full object-cover" 
          onError={() => setImgError(true)}
        />
      )}
      
      {/* Fallback: Initialen oder Platzhalter */}
      {(!src || imgError) && !customComponent && (
        <div className={`h-full w-full flex items-center justify-center text-white ${getBackgroundColor()}`}>
          {name ? getInitials() : (
            <svg className="h-1/2 w-1/2 text-gray-300 dark:text-gray-500" fill="currentColor" viewBox="0 0 20 20">
              <path fillRule="evenodd" d="M10 9a3 3 0 100-6 3 3 0 000 6zm-7 9a7 7 0 1114 0H3z" clipRule="evenodd" />
            </svg>
          )}
        </div>
      )}
      
      {/* Status-Indikator */}
      {status && (
        <span className={`absolute bottom-0 right-0 block h-${size === 'xs' || size === 'sm' ? '2' : '3'} w-${size === 'xs' || size === 'sm' ? '2' : '3'} rounded-full ring-2 ring-white dark:ring-gray-800 ${statusClasses[status]}`} />
      )}
    </div>
  );
};

export default Avatar;
