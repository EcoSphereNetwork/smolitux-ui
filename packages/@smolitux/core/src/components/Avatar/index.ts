// packages/@smolitux/core/src/components/Avatar/index.ts
export { default, type AvatarProps } from './Avatar';
