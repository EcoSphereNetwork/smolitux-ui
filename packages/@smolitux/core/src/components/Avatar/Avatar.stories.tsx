import type { Meta, StoryObj } from '@storybook/react';
import { Avatar } from './Avatar';

const meta: Meta<typeof Avatar> = {
  title: 'Components/Avatar',
  component: Avatar,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
};

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {
    children: 'Avatar',
  },
};

export const CustomStyle: Story = {
  args: {
    children: 'Custom Avatar',
    className: 'custom-style',
  },
};

export const Interactive: Story = {
  args: {
    children: 'Interactive Avatar',
    onClick: () => alert('Clicked!'),
  },
};
