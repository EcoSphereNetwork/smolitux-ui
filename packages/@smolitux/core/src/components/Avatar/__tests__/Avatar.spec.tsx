import React from 'react';
import { render } from '@testing-library/react';
import { Avatar } from '../Avatar';

describe('Avatar Snapshots', () => {
  it('renders with image correctly', () => {
    const { asFragment } = render(
      <Avatar 
        src="https://example.com/avatar.jpg" 
        alt="User Avatar" 
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders with initials correctly', () => {
    const { asFragment } = render(<Avatar name="John Doe" />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders with fallback icon correctly', () => {
    const { asFragment } = render(<Avatar />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders with different sizes correctly', () => {
    const { asFragment } = render(
      <>
        <Avatar size="xs" name="XS" />
        <Avatar size="sm" name="SM" />
        <Avatar size="md" name="MD" />
        <Avatar size="lg" name="LG" />
        <Avatar size="xl" name="XL" />
        <Avatar size="2.5rem" name="Custom" />
      </>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders with border correctly', () => {
    const { asFragment } = render(
      <Avatar name="John Doe" showBorder borderColor="blue-500" />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders with custom colors correctly', () => {
    const { asFragment } = render(
      <Avatar 
        name="John Doe" 
        bgColor="purple-500" 
        textColor="white" 
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders with badge correctly', () => {
    const { asFragment } = render(
      <>
        <Avatar name="Online" badge="online" />
        <Avatar name="Busy" badge="busy" />
        <Avatar name="Away" badge="away" />
        <Avatar name="Offline" badge="offline" />
        <Avatar name="Custom" badge="custom" badgeColor="pink-500" />
      </>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders with custom className correctly', () => {
    const { asFragment } = render(
      <Avatar name="John Doe" className="custom-avatar" />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders with square shape correctly', () => {
    const { asFragment } = render(
      <Avatar name="John Doe" shape="square" />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});