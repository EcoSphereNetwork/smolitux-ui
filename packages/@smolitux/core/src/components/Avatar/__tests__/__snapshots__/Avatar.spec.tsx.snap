// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar Snapshots renders with badge correctly 1`] = `
<DocumentFragment>
  <div
    badge="online"
    class="relative inline-flex items-center justify-center overflow-hidden h-12 w-12 text-base rounded-full  bg-gray-200 dark:bg-gray-700  "
  >
    <div
      class="h-full w-full flex items-center justify-center text-white bg-pink-500"
    >
      O
    </div>
  </div>
  <div
    badge="busy"
    class="relative inline-flex items-center justify-center overflow-hidden h-12 w-12 text-base rounded-full  bg-gray-200 dark:bg-gray-700  "
  >
    <div
      class="h-full w-full flex items-center justify-center text-white bg-yellow-500"
    >
      B
    </div>
  </div>
  <div
    badge="away"
    class="relative inline-flex items-center justify-center overflow-hidden h-12 w-12 text-base rounded-full  bg-gray-200 dark:bg-gray-700  "
  >
    <div
      class="h-full w-full flex items-center justify-center text-white bg-green-500"
    >
      A
    </div>
  </div>
  <div
    badge="offline"
    class="relative inline-flex items-center justify-center overflow-hidden h-12 w-12 text-base rounded-full  bg-gray-200 dark:bg-gray-700  "
  >
    <div
      class="h-full w-full flex items-center justify-center text-white bg-yellow-500"
    >
      O
    </div>
  </div>
  <div
    badge="custom"
    badgecolor="pink-500"
    class="relative inline-flex items-center justify-center overflow-hidden h-12 w-12 text-base rounded-full  bg-gray-200 dark:bg-gray-700  "
  >
    <div
      class="h-full w-full flex items-center justify-center text-white bg-yellow-500"
    >
      C
    </div>
  </div>
</DocumentFragment>
`;

exports[`Avatar Snapshots renders with border correctly 1`] = `
<DocumentFragment>
  <div
    class="relative inline-flex items-center justify-center overflow-hidden h-12 w-12 text-base rounded-full  bg-gray-200 dark:bg-gray-700  "
  >
    <div
      class="h-full w-full flex items-center justify-center text-white bg-teal-500"
    >
      JD
    </div>
  </div>
</DocumentFragment>
`;

exports[`Avatar Snapshots renders with custom className correctly 1`] = `
<DocumentFragment>
  <div
    class="relative inline-flex items-center justify-center overflow-hidden h-12 w-12 text-base rounded-full  bg-gray-200 dark:bg-gray-700  custom-avatar"
  >
    <div
      class="h-full w-full flex items-center justify-center text-white bg-teal-500"
    >
      JD
    </div>
  </div>
</DocumentFragment>
`;

exports[`Avatar Snapshots renders with custom colors correctly 1`] = `
<DocumentFragment>
  <div
    bgcolor="purple-500"
    class="relative inline-flex items-center justify-center overflow-hidden h-12 w-12 text-base rounded-full  bg-gray-200 dark:bg-gray-700  "
    textcolor="white"
  >
    <div
      class="h-full w-full flex items-center justify-center text-white bg-teal-500"
    >
      JD
    </div>
  </div>
</DocumentFragment>
`;

exports[`Avatar Snapshots renders with different sizes correctly 1`] = `
<DocumentFragment>
  <div
    class="relative inline-flex items-center justify-center overflow-hidden h-6 w-6 text-xs rounded-full  bg-gray-200 dark:bg-gray-700  "
  >
    <div
      class="h-full w-full flex items-center justify-center text-white bg-yellow-500"
    >
      X
    </div>
  </div>
  <div
    class="relative inline-flex items-center justify-center overflow-hidden h-8 w-8 text-sm rounded-full  bg-gray-200 dark:bg-gray-700  "
  >
    <div
      class="h-full w-full flex items-center justify-center text-white bg-red-500"
    >
      S
    </div>
  </div>
  <div
    class="relative inline-flex items-center justify-center overflow-hidden h-12 w-12 text-base rounded-full  bg-gray-200 dark:bg-gray-700  "
  >
    <div
      class="h-full w-full flex items-center justify-center text-white bg-blue-500"
    >
      M
    </div>
  </div>
  <div
    class="relative inline-flex items-center justify-center overflow-hidden h-16 w-16 text-lg rounded-full  bg-gray-200 dark:bg-gray-700  "
  >
    <div
      class="h-full w-full flex items-center justify-center text-white bg-yellow-500"
    >
      L
    </div>
  </div>
  <div
    class="relative inline-flex items-center justify-center overflow-hidden h-20 w-20 text-xl rounded-full  bg-gray-200 dark:bg-gray-700  "
  >
    <div
      class="h-full w-full flex items-center justify-center text-white bg-purple-500"
    >
      X
    </div>
  </div>
  <div
    class="relative inline-flex items-center justify-center overflow-hidden  rounded-full  bg-gray-200 dark:bg-gray-700  "
  >
    <div
      class="h-full w-full flex items-center justify-center text-white bg-yellow-500"
    >
      C
    </div>
  </div>
</DocumentFragment>
`;

exports[`Avatar Snapshots renders with fallback icon correctly 1`] = `
<DocumentFragment>
  <div
    class="relative inline-flex items-center justify-center overflow-hidden h-12 w-12 text-base rounded-full  bg-gray-200 dark:bg-gray-700  "
  >
    <div
      class="h-full w-full flex items-center justify-center text-white bg-gray-300 dark:bg-gray-600"
    >
      <svg
        class="h-1/2 w-1/2 text-gray-300 dark:text-gray-500"
        fill="currentColor"
        viewBox="0 0 20 20"
      >
        <path
          clip-rule="evenodd"
          d="M10 9a3 3 0 100-6 3 3 0 000 6zm-7 9a7 7 0 1114 0H3z"
          fill-rule="evenodd"
        />
      </svg>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Avatar Snapshots renders with image correctly 1`] = `
<DocumentFragment>
  <div
    class="relative inline-flex items-center justify-center overflow-hidden h-12 w-12 text-base rounded-full  bg-gray-200 dark:bg-gray-700  "
  >
    <img
      alt="User Avatar"
      class="h-full w-full object-cover"
      src="https://example.com/avatar.jpg"
    />
  </div>
</DocumentFragment>
`;

exports[`Avatar Snapshots renders with initials correctly 1`] = `
<DocumentFragment>
  <div
    class="relative inline-flex items-center justify-center overflow-hidden h-12 w-12 text-base rounded-full  bg-gray-200 dark:bg-gray-700  "
  >
    <div
      class="h-full w-full flex items-center justify-center text-white bg-teal-500"
    >
      JD
    </div>
  </div>
</DocumentFragment>
`;

exports[`Avatar Snapshots renders with square shape correctly 1`] = `
<DocumentFragment>
  <div
    class="relative inline-flex items-center justify-center overflow-hidden h-12 w-12 text-base rounded-none  bg-gray-200 dark:bg-gray-700  "
  >
    <div
      class="h-full w-full flex items-center justify-center text-white bg-teal-500"
    >
      JD
    </div>
  </div>
</DocumentFragment>
`;
